fastestimator/estimator.py

```diff
@@ -216,27 +229,59 @@ def _start_test(self):
+        loader = iter(self._configure_loader(self.pipeline.get_loader(self.system.mode, self.system.epoch_idx)))
+        network_epoch_ops = self.network.load_epoch(mode=self.system.mode, epoch=self.system.epoch_idx)
+        self.system.batch_idx = 0
+        while True:
+            try:
+                with Suppressor():
+                    batch = next(loader)
+                if self.system.batch_idx == self.system.max_steps_per_epoch and self.system.mode == "train":
+                    break
+                batch = self._configure_tensor(loader, batch)
+                effective_batch = self.network.get_effective_batch_input(batch, self.system.mode)
+                self._run_traces_on_batch_begin()
+                strategy = tf.distribute.get_strategy()
+                if isinstance(strategy, tf.distribute.MirroredStrategy):
+                    prediction = strategy.experimental_run_v2(
+                        self.network.forward_step_static,
+                        args=(effective_batch,
+                              state,
+                              network_epoch_ops,
+                              to_list(self.network.effective_outputs[self.system.mode])))
+                    batch = per_replica_to_global(batch)
+                    prediction = per_replica_to_global(prediction)
+                else:
+                    prediction = self.network.forward_step_static(
+                        effective_batch,
+                        state,
+                        network_epoch_ops,
+                        to_list(self.network.effective_outputs[self.system.mode]))
+                self._run_traces_on_batch_end(batch, prediction)
+                if self.system.mode == "train":
+                    self.system.update_global_step()
+                self.system.batch_idx += 1
+            except StopIteration:
                 break
-            batch = self._configure_tensor(loader, batch)
-            self._run_traces_on_batch_begin()
-            prediction = self.network.run_step(batch, {"mode": self.system.mode, "warmup": False})
-            self._run_traces_on_batch_end(batch, prediction)
-            if self.system.mode == "train":
-                self.system.update_global_step()
         self.network.unload_epoch()
         self._run_traces_on_epoch_end()
 
     def _configure_loader(self, loader):
         new_loader = loader
-        if isinstance(loader, DataLoader) and isinstance(self.network, TFNetwork):
-            data_type = to_type(to_tensor(loader.dataset[0], target_type="tensorflow"))
-            new_loader = tf.data.Dataset.from_generator(lambda: loader, data_type)
+        if isinstance(new_loader, DataLoader) and isinstance(self.network, TFNetwork):
+            batch = to_tensor(loader.dataset[0], target_type="tensorflow")
+            data_type = to_type(batch)
+            data_shape = to_shape(batch)
+            for key, shape in data_shape.items():
```

Comment:
is this reasonable for nested data?

Reply:
yes, the generator needs it :) 

Once we move out of the generator for multi-gpu, we can get rid of it

Reply:
i guess my question is more: will this work for nested data? if i'm reading these 2 lines properly we're figuring out the real shape but then replacing it with a list of None ?

Reply:
yes it will work on nested data.

we need the shape to be None because the shape between batch may change, but the rank won't (unless that's our assumption)

Reply:
but doesn't the following just throw out all of the actual shape information except for batch size?

data_shape[key] = [None] * (len(shape) + 1)

Reply:
right.  we are not using the full potential of `get_shape` function,    but I just don't want to implement a `get_rank` because it will be redundant.

Reply:
what if batch is as follows:

[ {"x": {"a":[5,10,15], "b":[1,2,3,4,5]}}, ... ]

then data_shape["x"] = [None, None, None] which doesn't seem useful

Reply:
interesting... here we go `get_rank`